### PR TITLE
Improve Adder location for touch screen devices

### PR DESCRIPTION
**Pull request by LMS007**
- Ensure that the adder location is always below the highlighted selection on a touch devices so that it does not compete with the copy/paste bar.
- Add a little extra margin (10px) to the adder on touch devices to give room for the selection handles
- Add helper method `isTouchDevice` for touch devices
- Add comment in isSelectionBackwards to explain that iOS does not have the same selection object as non-iOS devices
- Add 1 missing test coverage case in adder.js

------

fixes https://github.com/hypothesis/client/issues/2626

This ticket was initially created to address issues with the events dealing with drag and drop on a touch device.  Ideally, it would be nice if there were events that gave a "end" to a "drag" event, but no such event exists. I presume that is because you can re-adjust the selection handles after the fact. For all intents and purposes we're no longer going to try and hide the adder when a selection is being made.

That leads the second set of problems outlined in the ticket: The adder on touch devices tends to get in the way of the copy & paste bar when you make a selection from bottom to top (backwards) as you can see here.

(master branch)
102418112-f8270480-3fb1-11eb-9ba6-992729f9e60b

More so, I discovered that android and iOS are not equal in this behavoir. For reasons I don't yet understand, window.getSelection() does not return the same object on iOS that is returned on safari or chrome on mac desktop or android. In other words, it appears iOS is the outlier. For that reason, the condition in the last line of `isSelectionBackwards` always returns false in iOS because those two nodes are never the same. 

For that reason, iOS will never render the add near the start of a range selection and so the adder is almost always below the highlighted text. The only times it flips is when it's close to the bottom of the screen. 

With android, its different and will in fact render above the highlight selection if you create it backwards. This leads to problems because it cover the  build-in copy bar. This fix presented here is to force the adder to only render below a selection on all touch devices (except when off the bottom of screen). This is already the case on iOS (But I suspect it was inadvertent). This PR further forces both iOS and android (or rather, all touch devices) to render it below the highlighted text.

Additionally, I added a 10px margin to create separation between the adder the the highlighted text to give more room for the grab handles. These blue handles can impede the adder esp. on android where they are large.

Per best practice, I avoid using being specific with the user-agent. It's my hope we don't get into specific conditions based on iOS vs android, but I think it's very reasonable to have a condition for touch devices in general. 

In this next gif  you can see the adder never blocks the copy & paste bar.

(this branch)
android-adder

The adder can still render above a selection if it falls off the bottom of the screen! In this case, it will render under the copy and paste bar, but.

1. We can't know for sure where the copy & past bar is. In some cases it renders below when the selection is made near the top of the screen so we're always guessing in a way.
2. It's totally reasonable to assume users know how to, and can easily scroll to center the text better in the viewport. Scrolling on mobile is just part of the nature of the device and its reasonable to assume users can make scrolling corrections in these edge cases.

There is no way to guess where the copy & paste bar shows up, but these changes here fix the bulk of the problem. The remaining edge cases only show up when near the bottom or top of the screen.

### CODE REVIEW

File: src/annotator/test/adder-test.js [hunk 1]
```diff
@@ -199,6 +204,12 @@ describe('Adder', () => {
       assert.equal(target.arrowDirection, ARROW_POINTING_UP);
     });
 
+    it('does not position the adder above the top of the viewport even when selection is backwards', () => {
```
Comment (LMS007) [inline src/annotator/test/adder-test.js:207]: just adding a test for prexisting missing codecov

**Comment by esanzgar** [comment]
@LMS007 Do you know if Android's built-in copy bar always appears on the top of the selected text? For all languages? In all situations?

**Comment by robertknight** [comment]
I've just read through the PR description. The proposed approach generally sounds good to me. I was confused by this sentence:

> For reasons I don't yet understand, document.getSelection() does not return the same object on iOS that is returned on safari or chrome on mac desktop or android.

There is no `document.getSelection()` API. I guess you mean [window.getSelection()](https://developer.mozilla.org/en-US/docs/Web/API/Window/getSelection)? Can you clarify what you mean by "does not return the same object"? The API directly returns a `Selection` instance.

File: src/annotator/range-util.js [hunk 2]
```diff
@@ -10,6 +10,7 @@ export function isSelectionBackwards(selection) {
   }
 
   const range = selection.getRangeAt(0);
+  // Does not work in iOS because these two nodes are never equal.
   return range.startContainer === selection.focusNode;
```
Comment (robertknight) [inline src/annotator/range-util.js:14]: What was the relation between `range.startContainer` and `selection.focusNode` in your iOS testing? According to [the spec](https://w3c.github.io/selection-api/#dfn-focus):

> Each selections also have an anchor and a focus. If the selection's range is null, its anchor and focus are both null. If the selection's range is not null and its direction is forwards, its anchor is the range's start, and its focus is the end. Otherwise, its focus is the start and its anchor is the end.

So it seems from a reading of the spec that this check should be valid and if not, might indicate a bug that we should report.

File: src/annotator/adder.js [hunk 3]
```diff
@@ -180,9 +181,11 @@ export class Adder {
     // Set the initial arrow direction based on whether the selection was made
     // forwards/upwards or downwards/backwards.
     /** @type {ArrowDirection} */ let arrowDirection;
-    if (isRTLselection) {
+    if (isRTLselection && !isTouchDevice()) {
       arrowDirection = ARROW_POINTING_DOWN;
     } else {
+      // Render the adder below the selection for touch devices due to competing
+      // space with the copy/paste bar.
```
Comment (robertknight) [inline src/annotator/adder.js:188]: I think this sentence could be clarified to explain that "the copy/paste bar" is a native UI which typically appears above the selection on touch devices.

File: src/annotator/adder.js [hunk 4]
```diff
@@ -192,6 +195,9 @@ export class Adder {
     // and close to the end.
     const hMargin = Math.min(ARROW_H_MARGIN, selectionRect.width);
     const adderWidth = this._width();
+    // Render the adder a little lower on a touch device to provide room for the
+    // selection grab handles.
```
Comment (robertknight) [inline src/annotator/adder.js:199]: By this comment you mean that we're moving the adder lower so that interactions with it are not confused with interactions with the selection grab handles?

File: src/annotator/adder.js [hunk 5]
```diff
@@ -192,6 +195,9 @@ export class Adder {
...
+    const touchScreenMargin = isTouchDevice() ? 10 : 0;
```
Comment (robertknight) [inline src/annotator/adder.js:200]:
```suggestion
    const touchScreenOffset = isTouchDevice() ? 10 : 0;
```

"Offset" seems to me like a more appropriate term here than margin since we're talking about a position (the Y coordinate of the top of the adder) rather than a box (which has a margin around it).

**Review by robertknight — CHANGES_REQUESTED** [review]
The proposed UI changes sound good to me - thank-you for the detailed PR description. I had a query regarding the issue with `isSelectionBackwards` on iOS and also had a suggestion regarding how we test whether the device is a touch device or not, to make it consistent with how we detect in CSS.

**Comment by LMS007** [comment]
> @LMS007 Do you know if Android's built-in copy bar always appears on the top of the selected text? For all languages? In all situations?

I did not test other langs, thats a good question, but I do know it will render below if its near the top edge of the page, but how it calculates that is somewhat of a blackbox. I wouldn't want to guess and be wrong.

**Comment by esanzgar** [comment]
> I do know it will render below if its near the top edge of the page

In these situations where the built-in copy bar appears below the selection, will the built-in copy bar still obstruct the view of the adder?

File: src/annotator/test/adder-test.js [hunk 6]
```diff
@@ -221,6 +232,23 @@ describe('Adder', () => {
...
+            isTouchDevice: sinon.stub().returns(true),
+          },
+        });
+        const target = adderCtrl._calculateTarget(
+          rect(100, 200, 100, 20),
+          true
+        );
+        assert.isAbove(target.top, 220);
+        assert.equal(target.arrowDirection, ARROW_POINTING_UP);
+        $imports.$restore();
```
[the remark this answers is not in the document]
Comment (LMS007) [reply src/annotator/test/adder-test.js:249]: A quick search in the code, I don't see any usage of params in $restore() so I'll move it to the top level

File: src/shared/user-agent.js [hunk 7]
```diff
@@ -35,3 +35,13 @@ export const isIOS = (
...
 };
+
+/**
+ * Returns true when the device is a touch device such
+ * as android or iOS.
+ *
+ * @param _document {Document} - Test seam
+ */
+export const isTouchDevice = (_document = document) => {
+  return 'ontouchend' in _document;
```
[the remark this answers is not in the document]
Comment (LMS007) [reply src/shared/user-agent.js:46]: Sure, also, I didnt think the file name made sense anymore as this is now a device type detection, but I didn't have a better idea.

File: src/annotator/range-util.js (hunk 2 above)
Comment (LMS007) [reply src/annotator/range-util.js:14]: I need to get the exact objects so I can be more specific, its a bit hard to debug on my ipad, I need to find a better way.

File: src/annotator/adder.js (hunk 4 above)
Comment (LMS007) [reply src/annotator/adder.js:199]: correct. I will re-word that a bit

File: src/annotator/adder.js (hunk 5 above)
Comment (LMS007) [reply src/annotator/adder.js:200]: agreed

**Comment by LMS007** [comment]
> > I do know it will render below if its near the top edge of the page
> 
> In these situations where the built-in copy bar appears below the selection, will the built-in copy bar still obstruct the view of the adder?

Yes it will and with that, this solution is mostly fixing ~80% of the problem. My justification for ignoring these edge cases are simply:

- its somewhat more rare.
- I would expect users to generally be able to scroll on a touch device and center the selecting text they are trying to annotate. 

Having said that, there are 2 edge cases and both require us to guess when the copy and paste bar renders in a different place.

1. At the bottom when we render the adder above the selection -- on touch devices we could perhaps add a  lot of offset and to force the adder to be above the presumed copy/paste bar. Something like ~80px maybe required. 
2.  At the top, we do the same thing, but push it lower by that amount. 

We also (very likely) have to find a best fix model for both ios and android. I suggest we consider doing this if we still continue to see trouble on touch devices. We could land a follow up PR after this one to flush out these edge cases as best as we can.

**Comment by LMS007** [comment]
> I've just read through the PR description. The proposed approach generally sounds good to me. I was confused by this sentence:
> 
> > For reasons I don't yet understand, document.getSelection() does not return the same object on iOS that is returned on safari or chrome on mac desktop or android.
> 
> There is no `document.getSelection()` API. I guess you mean [window.getSelection()](https://developer.mozilla.org/en-US/docs/Web/API/Window/getSelection)? Can you clarify what you mean by "does not return the same object"? The API directly returns a `Selection` instance.

yes, window.getSelection, i typed it wrong. I will pull details between iOS and non-iOS and report back.

File: src/shared/test/user-agent-test.js [hunk 8]
```diff
@@ -19,6 +19,14 @@ describe('shared/user-agent', () => {
     });
   });
 
+  describe('isTouchDevice', () => {
+    it('calls `window.matchMedia` with the query string "(pointer: coarse)"', () => {
+      const matchMedia = sinon.spy(window, 'matchMedia');
```
Comment (robertknight) [inline src/shared/test/user-agent-test.js:24]: This spy is not reset anywhere. I would suggest to stub `matchMedia` in a `beforeEach` and reset in an `afterEach` block.

File: src/shared/user-agent.js [hunk 9]
```diff
@@ -35,3 +35,13 @@ export const isIOS = (
...
 };
+
+/**
+ * Returns true when the device is a touch device such
+ * as android or iOS.
+ *
+ * @param _window {Window} - Test seam
+ */
+export const isTouchDevice = (_window = window) => {
+  return _window.matchMedia('(pointer: coarse)').matches;
```
Comment (robertknight) [inline src/shared/user-agent.js:46]: From https://developer.mozilla.org/en-US/docs/Web/CSS/@media/pointer#browser_compatibility the only browser that is theoretically supported but may not support `(pointer: coarse)` is Firefox < 64. These devices will simply report as non-touch, which is fine. The same will happen with CSS rules that use this query.

File: src/annotator/range-util.js (hunk 2 above)
Comment (robertknight) [reply src/annotator/range-util.js:14]: This comment was marked as resolved but I didn't get an answer on the query above. Did you figure out what happens here on iOS?

**Review by robertknight — CHANGES_REQUESTED** [review]
Hi Kyle,

The changes look good except for a sinon stub that is not being reset. Also, did you figure out what was happening with `isSelectionBackwards` on iOS? What are the values of `range.startContainer` and `selection.focusNode` when the selection is backwards?

File: src/shared/test/user-agent-test.js (hunk 8 above)
Comment (LMS007) [reply src/shared/test/user-agent-test.js:24]: oops, good catch

File: src/shared/user-agent.js (hunk 9 above)
Comment (LMS007) [reply src/shared/user-agent.js:46]: I will add that link in a comment.. I think its valuable.

File: src/annotator/range-util.js [hunk 10]
```diff
@@ -10,6 +10,8 @@ export function isSelectionBackwards(selection) {
   }
 
   const range = selection.getRangeAt(0);
+  // Does not work correctly on iOS when selecting nodes backwards.
+  // https://bugs.webkit.org/show_bug.cgi?id=220523
```
Comment (robertknight) [inline src/annotator/range-util.js:14]: Thanks for filing this bug report. I notice that it links to a URL in an S3 bucket that you host. Just in case you remove this bucket in future for any reason I would suggest adding a copy of the HTML file as an attachment to the bug report. Also it may be worth mentioning the URL of this PR as additional context for anyone looking at the issue in future.

**Review by robertknight — APPROVED** [review]
LGTM. Sorry for the delay in re-reviewing this Kyle. Please do feel free to chase me up if you're waiting on a re-review that I've missed. I added a small request regarding the bug report that you filed upstream with WebKit.

File: src/annotator/range-util.js (hunk 10 above)
Comment (LMS007) [reply src/annotator/range-util.js:14]: done!
